stateful-autonomous/robot.py

```diff
@@ -6,7 +6,7 @@
 from robotpy_ext.autonomous import AutonomousModeSelector
 
 
-class MyRobot(wpilib.IterativeRobot):
+class MyRobot(wpilib.TimedRobot):
```

Comment:
I'm pretty sure this will break disastrously… are you sure the timing of the control loop still works as expected?

Reply:
It works in the sim.  The only issue is that the motors need to be invoked at 0 to avoid a motor timeout.  Oddly enough, Iterative did still work with no errors, but I saw this on the list to be updated so I tried to include it.

Reply:
By "it works", did you time the delay between the control loop iterations?

Reply:
I did not.  I am not certain I know how.  It seems once it enters a function, it will not do anything else in the main loop until it finishes.  So, either this proves the issue, or I cannot figure out a way to time the control loop.  I ran the different state machine options, and other than the motor timeout, they functioned.  These routines are all simple and sequential.  It is entirely possible that a more complex routine would break something. I could go back to the drawing board on this one, or at least test it with a real robot or some more complicated states.

Reply:
Hence why I'm fairly certain that this will break the control loop timing. You can reference the timed example, or I believe something was added into robotpy-wpilib-utilities to time the control loop as well based on it. Try running this in autonomous for a bit, then switch to teleop and see how it affects the timing.

Reply:
It took me a bit, but I was able to test the timing.  Originally, I was just looking at elapsed time in the timer, and that had too much operator error involved to be valid as it relied entirely upon me moving out of teleop at prices time.s repeatedly.  So, as suggested, I used the timer example and looked into loops/second.  This was much more informative.  

The first thing I noticed was both IterativeRobot and TimedRobot seem to treat the states as almost blocking calls.  Once auto enters the state machine, that seems to be all it does, and if motors are not explicitly addressed, there is a motor safety warning.  The exception is if the system moves out of auto at any time, the simulated robot will proceed as expected.

The second thing I noticed was, running teleop after auto (regardless of the state, or time in auto) resulted in consistent results.  The first is that IterativeRobot is always 30 loops/sec and TimedRobot is consistently between 49 and 51.  

The latter is the same result as the timer example.
So, my initial conclusion is that though there may be a better way to go about this, it seems the change from IterativeRobot to TimedRobot may actually work. 
Any thoughts about how to proceed?  I would be happy to revert the changes and only update the code to allow it to run on IterativeRobot if you think it would be better, or I could try some other tests.  I still have not had the chance to try on an actual robot, but could do so tomorrow (perhaps I should test the e-stop at least).
Sorry for the long post.

Reply:
Honestly, I'd rather get rid of the stateful autonomous stuff anyways, it's a precursor to magicbot and I'd rather people try that instead. Both iterative and timed robot are going to have similar timing related issues with StatefulAutonomous as it was designed with SampleRobot in mind.